staging/src/k8s.io/component-base/metrics/wrappers.go

```diff
@@ -65,6 +68,94 @@ type GaugeMetric interface {
 	SetToCurrentTime()
 }
 
+// PreContextGaugeMetric binds a Context to make a GaugeMetric
+type PreContextGaugeMetric interface {
+	// WithContext creates a GaugeMetric that uses the given Context
+	WithContext(ctx context.Context) GaugeMetric
+}
+
+// PreContextAndRegisterableGaugeMetric extends GaugeMetric with registration, collection, and context binding
+type PreContextAndRegisterableCollectorGaugeMetric interface {
```

Comment:
> The fact that these two things implement the same interface is an important insight from [@beorn7](https://github.com/beorn7) , and I think the code should clearly say that.

I don't think it convinces me to actually buy those hard to follow interfaces.
What really matters in my opinion is the fact that TimingHistogram has exactly the same interface as GaugeMetric.
The fact of registerable, kubeCollector etc. are implementation details.

So I agree that we need conditions like:
```
var _ GaugeMetric = &TimingHistogram{}
var _ GaugetVecMetric = &TimingHistogramVec{}
```

and the need to GaugeVecMetric interface below, but I don't buy the need for any other interface here.
They introduce more congnitive burden than gain.